aten/src/ATen/native/cpu/IndexKernel.cpp

```diff
@@ -119,10 +119,158 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
   });
 }
 
+static inline int64_t ensure_nonempty_dim(int64_t dim) {
+  return std::max<int64_t>(dim, 1);
+}
+
+static inline int64_t ensure_nonempty_size(const Tensor& t, int64_t dim) {
+  return t.dim() == 0 ? 1 : t.size(dim);
+}
+
+static inline int64_t ensure_nonempty_stride(const Tensor& t, int64_t dim) {
+  return t.dim() == 0 ? 1 : t.stride(dim);
+}
+
+using IdxVec = std::vector<int64_t>;
+static inline IdxVec ensure_nonempty_vec(IdxVec vec) {
+  if (vec.size() == 0) {
+    vec.push_back(1);
+  }
+  return vec;
+}
+
+static Tensor restride_dim(
+  const Tensor& src, int64_t dim,
+  IntArrayRef replacement_shape
+) {
+  auto strides = ensure_nonempty_vec(src.strides().vec());
+  strides[dim] = 0;
+  return src.as_strided(replacement_shape, strides);
+}
+
+static TensorIterator make_scatter_iterator(
+  Tensor& self, int64_t dim,
+  const Tensor& index, const Tensor& src
+) {
+  auto index_sizes = ensure_nonempty_vec(index.sizes().vec());
+  auto index_strides = ensure_nonempty_vec(index.strides().vec());
+
+  index_sizes[dim] = 1;
+  index_strides[dim] = 0;
+
+  auto self_restrided = restride_dim(self, dim, index_sizes);
+  auto index_restrided = index.as_strided(index_sizes, index_strides);
+  auto src_restrided = src.as_strided(index_sizes, index_strides);
+
+  auto iter = TensorIterator();
+  iter.dont_compute_common_dtype();
+  iter.dont_resize_outputs();
+  iter.add_output(self_restrided);
+  iter.add_input(src_restrided, src.device(), src.scalar_type());
+  iter.add_input(index_restrided);
+  iter.build();
+  return iter;
+}
+
+template <typename func_t>
+void scatter_base_cpu_kernel(
+  Tensor& self, int64_t dim,
+  const Tensor& index, const Tensor& src,
+  const std::string& method_name,
+  const func_t& f,
+  bool serial_exec = true
+) {
+  // Tests:
+  //  1. index.size(d) <= src.size(d) for all d
+  //  2. index.size(d) <= self.size(d) for all d != dim
+  bool is_wrong_shape = false;
+  int64_t self_dims = ensure_nonempty_dim(self.dim());
+  for (int64_t d = 0; d < self_dims; ++d) {
+    int64_t index_d_size = ensure_nonempty_size(index, d);
+    if (index_d_size > ensure_nonempty_size(src, d)) {
+      is_wrong_shape = true;
+      break;
+    }
+    if (d != dim) {
+      if (index_d_size > ensure_nonempty_size(self, d)) {
+        is_wrong_shape = true;
+        break;
+      }
+    }
+  }
+  TORCH_CHECK(!is_wrong_shape,
+    "Expected index [", index.sizes(), "]"
+    " to be smaller size than src [", src.sizes(), "]"
+    " and to be smaller than self [", self.sizes(), "]"
+    " apart from dimension ", dim
+  );
+
+  DimVector indexed_sizes;
+  DimVector indexed_strides;
+  indexed_sizes.push_back(ensure_nonempty_size(self, dim));
+  indexed_strides.push_back(
+    1//ensure_nonempty_stride(self, dim) * self.element_size()
```

Comment:
Can you please check my understanding of what's going on here? 
It looks like the reason you are getting a speed-up compared to my original implementation is that you are bypassing offset calculation in the cpu_index_kernel, because the `f` lambda you are passing to cpu index kernel does not require offset and you are relying on the compiler to optimizer `offset` computation away. But then 
it looks like you don't need `indexed_sizes` and `indexed_strides` at all, other than as unused arguments to cpu_index_kernel - in the cpu_index_kernel they are used to construct Indexer object, but you bypass Indexer object altogether? If that's the case, there's no need to initialize them to sensible values, and also it would be nice to have a comment about the way you are using cpu_index_kernel because it is non-trivial.

Reply:
Yes, you are right, this is way I asked you whether there is some other kernel which only returns pointers...